netket/experimental/operator/_fermion_operator_2nd_jax.py

```diff
@@ -0,0 +1,686 @@
+    not_zero = ~zero
+    w_final = weight * not_zero * sign
+    # return the xp, the mel and wether mel is zero
+    return x_final.astype(x.dtype), w_final, not_zero
+
+
+@partial(jax.vmap, in_axes=(None, 0, 0, 0, None), out_axes=(-2, -1, -1))
+def _apply_terms_scan(x, w, sites, daggers, unroll):
+    return _apply_term_scan(x, w, sites, daggers, unroll=unroll)
+
+
+@partial(jax.jit, static_argnums=4)
+def apply_terms_scan(x, w, sites, daggers, unroll=1):
+    return _apply_terms_scan(x, w, sites, daggers, unroll)
+
+
+def _apply_term_scan_math(x, weight, sites, daggers, unroll=1):
+    if not (
+        jnp.issubdtype(x.dtype, jnp.signedinteger)
+        or jnp.issubdtype(x.dtype, jnp.floating)
+    ):
+        raise ValueError(
+            f"incompatible dtype for x, expect a type with sign but got {x.dtype}"
+        )
+
+    if len(sites) == 0:  # constant diagonal term
+        return x, jnp.full(x.shape[:-1], weight), jnp.full(x.shape[:-1], True)
+
+    daggers = daggers.astype(x.dtype)
+
+    _and = jax.lax.mul
+    # here it's not really xor anymore,
+    # as we take compute prod of the sign directly
+    sgn = jnp.ones(x.shape[:-1], dtype=x.dtype)
+    _sign_preprocess = lambda x: 2 * x - 1
+    _xor = jax.lax.mul
+    _reduce_xor = lambda x, axes: jax.lax.reduce(x, 1, _xor, axes)
+    _sign_postprocess = lambda x: x
+
+    # use max for or to avoid overflow
+    _or = jax.lax.max
+    _not = lambda x: 1 - x
+    _equal = jax.lax.eq
+
+    zero = jnp.zeros(x.shape[:-1], dtype=x.dtype)
+    init = x, sgn, zero
+    xs = sites, daggers
+
+    def f(carry, xs):
+        site, dagger = xs
+        x_, sgn, zero = carry
+
+        x_new = x_.at[..., site].set(dagger)
+
+        n_orbitals = x_.shape[-1]
+        mask_all_up_to_site = (jnp.arange(n_orbitals, dtype=sites.dtype) < site).astype(
+            x_.dtype
+        )
+        x_masked = _and(x_, mask_all_up_to_site[None])
+        sgn = _xor(sgn, _reduce_xor(_sign_preprocess(x_masked), (x_.ndim - 1,)))
+
+        zero = _or(zero, _equal(x_.at[..., site].get(), dagger).astype(x_.dtype))
+
+        return (x_new, sgn, zero), None
+
+    (x_final, sgn, zero), _ = jax.lax.scan(f, init, xs, unroll=unroll)
+
+    sign = _sign_postprocess(sgn).astype(x.dtype)
+    not_zero = _not(zero).astype(x.dtype)
+    # cast to avoid unwated type promotion
+    w_final = weight * (not_zero * sign).astype(weight.dtype)
+    # return the xp, the mel and wether mel is zero
+    return x_final, w_final, not_zero
+
+
+@partial(jax.vmap, in_axes=(None, 0, 0, 0, None), out_axes=(-2, -1, -1))
+def _apply_terms_scan_math(x, w, sites, daggers, unroll):
+    return _apply_term_scan_math(x, w, sites, daggers, unroll=unroll)
+
+
+@partial(jax.jit, static_argnums=(4, 5))
+def apply_terms_scan_math(x, w, sites, daggers, unroll=1, dtype=jnp.float32):
+    # TODO avoid casting here, return orig type then
+    x = x.astype(dtype)
+    return _apply_terms_scan_math(x, w, sites, daggers, unroll)
```

Comment:
```suggestion
```
this was an experiment I did for gpu (I had in mind of using half or even lower precision),
iirc it didn't perform much better, and we can probably remove this one, in favour of the other implementations.